src/Microsoft.Identity.Client/Azure/ManagedIdentity.cs

```diff
@@ -0,0 +1,370 @@
+        {
+            var requestUri = $"{_endpoint}?resource={resourceUri}&api-version={Constants.ManagedIdentityAppServiceApiVersion}";
+            var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+            request.Headers.Add("Secret", _secret);
+            return request;
+        }
+    }
+
+    /// <summary>
+    /// Used to hold the deserialized token response. 
+    /// </summary>
+    [DataContract]
+    internal class TokenResponse
+    {
+        private const string TokenResponseFormatExceptionMessage = "Token response is not in the expected format.";
+
+        internal enum DateFormat
+        {
+            Unix,
+            DateTimeString
+        };
+
+        // MSI endpoint return access_token
+        [DataMember(Name = "access_token", IsRequired = false)]
+        public string AccessToken { get; private set; }
+
+        // MSI endpoint return expires_on
+        [DataMember(Name = "expires_on", IsRequired = false)]
+        public string ExpiresOn { get; private set; }
+
+        [DataMember(Name = "error_description", IsRequired = false)]
+        public string ErrorDescription { get; private set; }
+
+        // MSI endpoint return token_type
+        [DataMember(Name = "token_type", IsRequired = false)]
+        public string TokenType { get; private set; }
+
+        /// <summary>
+        /// Parse token response returned from OAuth provider.
+        /// While more fields are returned, we only need the access token. 
+        /// </summary>
+        /// <param name="tokenResponse">This is the response received from OAuth endpoint that has the access token in it.</param>
+        /// <returns></returns>
+        public static TokenResponse Parse(string tokenResponse)
+        {
+            try
+            {
+                return JsonHelper.DeserializeFromJson<TokenResponse>(Encoding.UTF8.GetBytes(tokenResponse));
+            }
+            catch (Exception exp)
+            {
+                throw new FormatException($"{TokenResponseFormatExceptionMessage} Exception Message: {exp.Message}");
+            }
+        }
+    }
+
+    /// <summary>
+    /// IManagedIdentityConfiguration provides the configurable properties for the ManagedIdentityProbe
+    /// </summary>
+    public interface IManagedIdentityConfiguration
+    {
+        /// <summary>
+        /// ManagedIdentitySecret is the secret for use in Azure AppService
+        /// </summary>
+        string ManagedIdentitySecret { get; }
+
+        /// <summary>
+        /// ManagedIdentityEndpoint is the AppService endpoint
+        /// </summary>
+        string ManagedIdentityEndpoint { get; }
+
+        /// <summary>
+        /// VMManagedIdentityEndpoint is the VM's default managed identity endpoint
+        /// </summary>
+        string VMManagedIdentityEndpoint { get; }
+
+        /// <summary>
+        /// ClientId is the user assigned managed identity for use in VM managed identity
```

Comment:
optional: in the case of user-assigned managed identity ?

Reply:
Yes, client id would be used here for user-assigned identities.